### DISCUSSION TITLE
WIRE TIMEOUTs

### DISCUSSION THREAD

**Discussion by neilh10** [neilh10/STC3100arduino#4] [General]
How to do WIRE TIMEOUTs - which should be an essential part of WIRE/I2C management.

The docs suggests "Code that needs to be portable across platforms and versions can use the WIRE_HAS_TIMEOUT macro, which is only defined when Wire.setWireTimeout(), Wire.getWireTimeoutFlag() and Wire.clearWireTimeout() are all available."
https://www.arduino.cc/reference/en/language/functions/communication/wire/setwiretimeout/
https://www.arduino.cc/reference/en/language/functions/communication/wire/getwiretimeoutflag/

For the compiler supplied by .platformio\packages\framework-arduino-avr there doesn't appear to a default define of  WIRE_HAS_TIMEOUT

in release 0.3.0 the label WIRE_HAS_TIMEOUT is defined as part of including ARDUINO_ARCH_STM32

I propose the code has  WIRE_HAS_TIMEOUT is defined by default - and if its needed to be removed, then the following should be added to the ARDUINO_ARCH_STM32

**Comment by fronders** [comment]
Okay I was surprised it's not defined, I then looked up the Wire lib source supplied by Arduino AVR core - turns out they did not add it. Documentation is there, but code is non, it's a bug on Arduino core side. 

This has been open for 3+ years 😬😬😬
https://github.com/arduino/ArduinoCore-avr/pull/362

I guess it got buried because of other changes, I'll open an issue just for that separately

**Comment by fronders** [comment]
Ok, I did some more research :) 

Turns out AVR core is [the only one](https://github.com/search?q=org%3Aarduino%20getWireTimeoutFlag&type=code) that has these timeout functions in Wire lib. Other arches would simply not compile if the code references `getWireTimeoutFlag()` etc

So since AVR's Wire is missing the `WIRE_HAS_TIMEOUT` in my opinion it's just easier to check for `ARDUINO_ARCH_AVR` instead and use the timeout functions if it's there:

```cpp
#if defined(ARDUINO_ARCH_AVR)
        if (_i2c->getWireTimeoutFlag()) {
            return TWOWIRE_ERR_TIMEOUT;
        }
#elif defined(ARDUINO_ARCH_STM32)
        I2C_HandleTypeDef * hi2c = _i2c->getHandle();
        if (hi2c && HAL_I2C_GetError(hi2c) != HAL_I2C_ERROR_NONE) {
            return TWOWIRE_ERR_TIMEOUT;
        }
#endif
```

Essentially in the merged code just replace `WIRE_HAS_TIMEOUT` with `ARDUINO_ARCH_AVR`.
Then it uses `getWireTimeoutFlag()` if it's AVR, `HAL_I2C_GetError()` is it's STM32 and nothing for other arches (since they don't have any of these functions anyway)

I can add a PR for that if you want

**Comment by fronders** [comment]
And I'd rather not define `WIRE_HAS_TIMEOUT` from inside the lib, since in the future if the AVR Wire implementation is fixed, it might be in conflict with that because or symbol redefinition

**Comment by neilh10** [comment]
Thanks for the research and suggestion. 
updated [commit 9802d79cb763c9d8ff6db86582c7710a6b32c455](https://github.com/neilh10/STC3100arduino/commit/9802d79cb763c9d8ff6db86582c7710a6b32c455)

**Comment by neilh10** [comment]
If there are no other comments I will tag it as a release.

**Reply by fronders** [reply]
Nope, I think everything's sorted out, thanks!
